File: app/src/main/java/closer/vlllage/com/closer/handler/phone/PhoneAdapter.java
package closer.vlllage.com.closer.handler.phone;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import closer.vlllage.com.closer.R;
import closer.vlllage.com.closer.api.models.ReactionResult;
import closer.vlllage.com.closer.handler.group.PhotoActivityTransitionHandler;
import closer.vlllage.com.closer.handler.helpers.PhotoHelper;
import closer.vlllage.com.closer.handler.helpers.Val;
import closer.vlllage.com.closer.pool.PoolMember;
import closer.vlllage.com.closer.pool.PoolRecyclerAdapter;

public class PhoneAdapter extends PoolRecyclerAdapter<PhoneAdapter.ViewHolder> {

    private final OnReactionClickListener onReactionClickListener;
    private List<ReactionResult> items = new ArrayList<>();

    public PhoneAdapter(PoolMember poolMember, OnReactionClickListener onReactionClickListener) {
        super(poolMember);
        this.onReactionClickListener = onReactionClickListener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        return new ViewHolder(LayoutInflater.from(viewGroup.getContext())
                .inflate(R.layout.phone_item, viewGroup, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder viewHolder, int position) {
        ReactionResult reaction = items.get(position);

        viewHolder.name.setText(reaction.phone.name);
        viewHolder.reaction.setText(reaction.reaction);

        if ($(Val.class).isEmpty(reaction.phone.photo)) {
            viewHolder.photo.setVisibility(View.GONE);
        } else {
            viewHolder.photo.setVisibility(View.VISIBLE);
            $(PhotoHelper.class).loadCircle(viewHolder.photo, reaction.phone.photo);
            viewHolder.photo.setOnClickListener(v -> {
                $(PhotoActivityTransitionHandler.class).show(viewHolder.photo, reaction.phone.photo);
            });
        }

        viewHolder.itemView.setOnClickListener(v -> {
            onReactionClickListener.onReactionClick(reaction);
        });
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public void setItems(List<ReactionResult> items) {
        this.items = items;
        notifyDataSetChanged();
    }

    public List<ReactionResult> getItems() {
        return items;
    }

    static class ViewHolder extends RecyclerView.ViewHolder {

        ImageView photo;
        TextView name;
        TextView reaction;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);

            photo = itemView.findViewById(R.id.photo);
            name = itemView.findViewById(R.id.name);
            reaction = itemView.findViewById(R.id.reaction);
        }
    }

    public interface OnReactionClickListener {
        void onReactionClick(ReactionResult  reaction);
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/phone/PhoneAdapter.kt
package closer.vlllage.com.closer.handler.phone

import android.content.res.ColorStateList
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import closer.vlllage.com.closer.R
import closer.vlllage.com.closer.api.models.ReactionResult
import closer.vlllage.com.closer.handler.group.PhotoActivityTransitionHandler
import closer.vlllage.com.closer.handler.helpers.PhotoHelper
import closer.vlllage.com.closer.pool.PoolRecyclerAdapter
import com.queatz.on.On

open class PhoneAdapter(on: On, private val onReactionClickListener: (ReactionResult) -> Unit) : PoolRecyclerAdapter<RecyclerView.ViewHolder>(on) {
    var items: List<ReactionResult> = listOf()
        set(items) {
            field = items
            isLoading = false
            notifyDataSetChanged()
        }

    var isLoading = true

    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return ViewHolder(LayoutInflater.from(viewGroup.context)
                .inflate(R.layout.phone_item, viewGroup, false))
    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        when (viewHolder) {
            is ViewHolder -> {
                val reaction = this.items[position]

                viewHolder.name.text = on<NameHandler>().getName(reaction.phone!!.id!!)
                viewHolder.reaction.text = reaction.reaction

                if (reaction.phone!!.photo == null) {
                    viewHolder.photo.setImageResource(R.drawable.ic_person_black_24dp)
                    viewHolder.photo.imageTintList = ColorStateList.valueOf(on<PhotoHelper>().colorForPhone(reaction.phone!!.id!!))
                    viewHolder.photo.scaleType = ImageView.ScaleType.CENTER_INSIDE
                } else {
                    viewHolder.photo.imageTintList = null
                    on<PhotoHelper>().loadCircle(viewHolder.photo, reaction.phone!!.photo!!)
                    viewHolder.photo.scaleType = ImageView.ScaleType.CENTER_CROP
                }
                viewHolder.photo.setOnClickListener { v ->
                    reaction.phone?.photo?.let { on<PhotoActivityTransitionHandler>().show(viewHolder.photo, it) } }

                viewHolder.itemView.setOnClickListener { v -> onReactionClickListener.invoke(reaction) }
            }
        }
    }

    override fun getItemCount() = items.size

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var photo: ImageView = itemView.findViewById(R.id.photo)
        var name: TextView = itemView.findViewById(R.id.name)
        var reaction: TextView = itemView.findViewById(R.id.reaction)
    }
}
